README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# knnwtsim

<!-- badges: start -->

[![R-CMD-check](https://github.com/mtrupiano1/knnwtsim/workflows/R-CMD-check/badge.svg)](https://github.com/mtrupiano1/knnwtsim/actions)
[![CRAN
status](https://www.r-pkg.org/badges/version/knnwtsim)](https://CRAN.R-project.org/package=knnwtsim)
<!-- badges: end -->

The goal of `knnwtsim` is to provide a package to share and implement a
forecasting methodology using k nearest neighbors (KNN) primarily for
situations where the response series of interest can be predicted by a
combination of its’ own recent realizations, its own periodic patterns,
and by the values of one or multiple exogenous predictors.

The functions of this package are focused primarily into two components.
The first being calculation of a similarity measure which takes into
account all three factors listed above (`S_w`), and can weight the
degree to which each component contributed to the overall similarity.
The second being the usage of this measure to identify neighbors and
perform KNN regression. For a more formal discussion of the methodology
used and the key user facing functions in `knnwtsim` please see Trupiano
(2021) \<arXiv:2112.06266v1>.

## Formulation

### Weighted Similarity Measure `S_w`

The Similarity measure I have formulated for incorporating the three
factors listed above is:

`S_w = alpha * S_t + beta * S_p + gamma * S_x`

Where `S_t` is the metric used to calculate the similarity matrix
generated by `StMatrixCalc` to measure similarity in terms of pure
recency between all observations. `alpha` is the weight between 0-1
assigned to this matrix in the calculation of `S_w`.

`S_p` is the metric used to calculate the similarity matrix generated by
`SpMatrixCalc` to measure similarity in terms of where each observation
falls along a periodic cycle relative to all others. `beta` is the
weight between 0-1 assigned to this matrix in the calculation of `S_w`.

`S_x` is the metric used to calculate the similarity matrix generated by
`SxMatrixCalc` to measure similarity between all observations in terms
of the values of one or more exogenous predictors associated with a
given observation. `gamma` is the weight between 0-1 assigned to this
matrix in the calculation of `S_w`.

The function `SwMatrixCalc` calls each of the previous similarity matrix
functions to generate the final matrix to use in `knn.forecast`.

The weights `alpha`, `beta`, and `gamma` are recommended to be set so
that they sum to 1. In this case each element of final similarity matrix
should also be between 0-1, with the diagonal elements being equal to 1.

#### Component Similarity Details

Initially `S_t`, `S_p`, and `S_x` are calculated as dissimilarities,
`D_t`, `D_p`, and `D_x`. They are then transformed to similarities by
the formula `1 / (D+1)`, where `D` is any dissimilarity measure. This
also ensures each element of the similarity matrices fall in the range
`(0,1]`, with 1 representing the greatest similarity, and the values
approaching 0 the least.

When `StMatrixCalc` generates the intermediate dissimilarity matrix
using `D_t`, it calls the function `TempAbsDissimilarity` for each
pairwise combination of the input vector. Which provided the time orders
of two points `y_i` and `y_j`, where the time orders are `i` and
`j`.Then `TempAbsDissimilarity(i, j)` will return the absolute
difference between the two, i.e `abs(i - j)`.

When `SpMatrixCalc` generates the intermediate dissimilarity matrix
using `D_p`, it calls the function `SeasonalAbsDissimilarity` for each
pairwise combination of the input vector. This function takes the values
corresponding to the periods of two points in an overall
periodic/seasonal cycle, and the total number of periods in a cycle.
Where we have two points `y_i` and `y_j` with `p_i` and `p_j`
representing the corresponding period of those points, and a total of
`p_max` periods in a full periodic cycle. The value returned by
`SeasonalAbsDissimilarity(p_i, p_j, p_max)` will be
`min(DirectDis, AroundDis)` where `DirectDis <- abs(p_i - p_j)` and
`AroundDis <- abs(min(p_i, p_j) - 1) + abs(p_max - max(p_i, p_j)) + 1`.

This formulation is based on the idea that generally the periods at the
very end and very beginning of a cycle should be fairly similar. To
clarify through an example, in a monthly cycle where 1 represents
January and 12 represents December, these two months are generally more
similar to each-other than either are to July at period 7. Continuing
with this monthly example, if we have three points which occur in
January, March, and November: `p_i = 1`, `p_j = 3`, `p_k = 11` with
`p_max = 12`. Then `SeasonalAbsDissimilarity(p_i, p_j, p_max) = 2`,
`SeasonalAbsDissimilarity(p_i, p_k, p_max) = 2`, and
`SeasonalAbsDissimilarity(p_j, p_k, p_max) = 4`.

Finally, `SxMatrixCalc` uses the `stats::dist` function to produce the
intermediate dissimilarity matrix using `D_x`, for an input matrix or
vector using the method indicated by the `XdistMetric` argument of
`SxMatrixCalc`. Naturally, this limits `D_x` to the methods available in
`stats::dist`.

### KNN Forecasting

K nearest neighbors forecasting is implemented in this package through
the function `knn.forecast`. Using a provided similarity matrix, which
is not required to be calculated using `S_w` specifically, the function
will perform K Nearest Neighbors regression on each point in a specified
index `f.index.in`, returning the mean of the identified, `k.in`,
neighbors in the response series `y.in`.

Mathematically the estimate for a given point `y_t` is formulated as the
mean of the previous points in the series `y_i` identified to be in the
neighborhood of `y_t`, `K(y_t)`.

In the neighborhood `K(y_t)` will be the `k` observations of `y_i` with
the highest similarity to `y_t` of all eligible members of the time
series, meaning `i < t`. Currently in `knn.forecast` this eligibility
constraint is enforced by only considering the rows of the similarity
matrix at indices which are not present in `f.index.in` when selecting
neighbors while performing KNN regression on each point in `f.index.in`,
thus preventing the corresponding points in `y.in` at those indices from
being selected as neighbors. Furthermore, rows and columns of
`Sim.Mat.in` at indices greater than the maximum value of `f.index.in`
will be removed.

## Installation

You can install the most recent stable version from CRAN with

``` r
install.packages("knnwtsim")
```

You can install the development version of knnwtsim from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("mtrupiano1/knnwtsim")
```

## Example with Known Similarity Matrix Weights and `k`

This is a basic example which shows a full forecasting workflow if the
weights to use in the generation of `S_w`, and the hyperparameter `k`
are known:

``` r
library(knnwtsim)
# Pull a series to forecast
data("simulation_master_list")
series.index <- 15
ex.series <- simulation_master_list[[series.index]]$series.lin.coef.chng.x

# Weights pre tuned by random search. In alpha, beta, gamma order
pre.tuned.wts <- c(0.2148058, 0.2899638, 0.4952303)
pre.tuned.k <- 5

df <- data.frame(ex.series)
# Generate vector of time orders
df$t <- c(1:nrow(df))

# Generate vector of periods
nperiods <- simulation_master_list[[series.index]]$seasonal.periods
df$p <- rep(1:nperiods, length.out = nrow(df))

# Pull corresponding exogenous predictor(s)
X <- as.matrix(simulation_master_list[[series.index]]$x.chng)
XdistMetric <- "euclidean"

# Number of points to set aside for validation
val.len <- ifelse(nperiods == 12, nperiods, nperiods * 2)

# Calculate the weighted similarity matrix using Sw
Sw.ex <- SwMatrixCalc( # For the recency similarity St
  t.in = df$t
  # For the periodic similarity Sp
  , p.in = df$p, nPeriods.in = nperiods
  # For the exogenous similarity Sx
  , X.in = X, XdistMetric.in = XdistMetric
  # Weights to be applied to each similarity
  , weights = pre.tuned.wts
)

# View the top corner of the weighted similarity matrix Sw
cat("\n Dimensions and Slice of S_w \n")
#> 
#>  Dimensions and Slice of S_w
print(dim(Sw.ex))
#> [1] 100 100
print(Sw.ex[1:5, 1:5])
#>           1         2         3         4         5
#> 1 0.9999999 0.5825337 0.4249690 0.4115130 0.5589633
#> 2 0.5825337 0.9999999 0.5989170 0.5690332 0.4841708
#> 3 0.4249690 0.5989170 0.9999999 0.6673498 0.4414853
#> 4 0.4115130 0.5690332 0.6673498 0.9999999 0.5582553
#> 5 0.5589633 0.4841708 0.4414853 0.5582553 0.9999999
```

``` r
# Index we want to forecast
val.index <- c((length(ex.series) - val.len + 1):length(ex.series))

# Generate the forecast
knn.frcst <- knn.forecast(
  Sim.Mat.in = Sw.ex,
  f.index.in = val.index,
  k.in = pre.tuned.k,
  y.in = ex.series
)
```

<img src="man/figures/README-knn-forecast-plot-1.png" width="100%" style="display: block; margin: auto;" />

## Example with Tuning of Similarity Matrix Weights

In most cases you will likely want to tune the hyperparameters used in
the construction of `S_w`, and the number of nearest neighbors, `k`, to
consider for any given point. There are many approaches that can be
taken to accomplish this tuning, and many users may choose to implement
their preferred approach. However, for those who want something
pre-built I have included a simple tuning function with the package
called `knn.forecast.randomsearch.tuning`. This function creates a
randomly generated `Grid` of potential sets of hyperparameters
`k`,`alpha`,`beta`,`gamma` and generates a forecast of length `test.h`
on the last `test.h` points in the series. The best set of parameters
are selected based on which row of the `Grid` led to the forecast with
the lowest MAPE (Mean Absolute Percent Error) over the points in the
forecast, and returned as the `weight.opt` and `k.opt` items of the list
object returned by the function. Additionally, the ‘optimal’ similarity
matrix to pass to `knn.forecast` is also returned as `Sw.opt`, as are
the testing grid (`Grid`), best test MAPE (`Test.MAPE.opt`), and the
full vector of test MAPE values (`MAPE.all`) corresponding to each row
in `Grid`.

If you do not want to tune the hyperparameters on the entire series and
prefer to leave some points for validation of the tuned result the
`val.holdout.len` will remove that many points from the end of the
series, and will perform the test forecasts on the `test.h` points at
the end of the series after the validation observations are removed.

Using the example series from the previous section, we can reproduce the
‘pre-tuned’ weights from the previous section and generate a forecast
using the included tuning function.

``` r
# Calculate component similarity matrices
St.ex <- StMatrixCalc(df$t)
Sp.ex <- SpMatrixCalc(df$p, nPeriods = nperiods)
Sx.ex <- SxMatrixCalc(X)

# Set seed for reproducibility
set.seed(10)
# Run tuning function
tuning.ex <- knn.forecast.randomsearch.tuning(
  grid.len = 10**4,
  y.in = ex.series,
  St.in = St.ex,
  Sp.in = Sp.ex,
  Sx.in = Sx.ex,
  test.h = val.len,
  max.k = NA,
  val.holdout.len = val.len
)


cat("\n Tuned Hyperparameters \n")
#> 
#>  Tuned Hyperparameters
cat("\n S_w Weights \n")
#> 
#>  S_w Weights
print(tuning.ex$weight.opt)
#> [1] 0.2148058 0.2899638 0.4952303
cat("\n k \n")
#> 
#>  k
print(tuning.ex$k.opt)
#> [1] 5

# Pull out tuned S_w and k
k.opt.ex <- tuning.ex$k.opt
Sw.opt.ex <- tuning.ex$Sw.opt
```

``` r
# Generate the forecast
knn.frcst.tuned <- knn.forecast(
  Sim.Mat.in = Sw.opt.ex,
  f.index.in = val.index,
  k.in = k.opt.ex,
  y.in = ex.series
)
```

<img src="man/figures/README-tuned-plot-1.png" width="100%" style="display: block; margin: auto;" />

## Forecasting with Prediction Intervals

The methodology used to produce forecasts with prediction intervals in
this package is based on the description of “Prediction intervals from
bootstrapped residuals” from chapter 5.5 of Hyndman R, Athanasopoulos G
(2021)
<https://otexts.com/fpp3/prediction-intervals.html#prediction-intervals-from-bootstrapped-residuals>,
modified as needed for use with KNN regression. The function
`knn.forecast.boot.intervals` is used to implement this method. The
algorithm employed starts by calculating a pool of forecast errors to
later sample from. If there are `n` points prior to the first
observation indicated in `f.index.in` then there will be `n - k.in`
errors generated by one-step ahead forecasts starting with the point of
the response series at the index `k.in + 1`. The first `k.in` points
cannot be estimated because a minimum of `k.in` eligible neighbors would
be needed. The optional `burn.in` argument can be used to increase the
number of points from the start of the series that need to be available
as neighbors before calculating errors for the pool. Next, `B` possible
paths the series could take are simulated using the pool of errors. Each
path is simulated by calling `knn.forecast`, estimating the first point
in `f.index.in`, adding a sampled forecast error, then adding this value
to the end of the series. This process is then repeated for the next
point in `f.index.in` until all have been estimated. The final output
interval estimates are calculated for each point in `f.index.in` by
taking the appropriate percentiles of the corresponding simulations of
that point. The output is returned as a list with the upper and lower
prediction interval bounds based on the supplied confidence level, as
well as the mean and median of the simulated points for each observation
in the forecast index. All `B` simulations can be returned if desired
using the optional `return.simulations = TRUE` argument. Below we have
an example, where we generate an interval forecast at the 95% confidence
level, and plot the results. Note that the mean forecast from this
function may differ from the point forecast generated by `knn.forecast`.
This difference is driven by the random sampling of residuals to
generate the simulated paths, as well as the behavior of adding
simulated points in a given path to the list of eligible neighbors when
repeatably calling `knn.forecast`.

``` r
# Produce interval forecast list
interval.forecast <- knn.forecast.boot.intervals(
  Sim.Mat.in = Sw.opt.ex,
  f.index.in = val.index,
  y.in = ex.series,
  k.in = k.opt.ex
)

# Pull out desired components
lb <- interval.forecast$lb
ub <- interval.forecast$ub
mean.boot <- interval.forecast$mean
```

<img src="man/figures/README-bootstrap-interval-plot-1.png" width="100%" style="display: block; margin: auto;" />


NEWS.md
# knnwtsim 1.0.0

## Major Changes
* The function `knn.forecast.boot.intervals()` has been added to provide the capability to generate interval forecasts and simulations with the package. 

* The default value of `max.k` in `knn.forecast.randomsearch.tuning()` has been changed from `NA` to `NULL`, to bring it in line with the approach I took to optional arguments elsewhere in the package. However, passing `max.k = NA` will still work, and behave in the same manner as `max.k = NULL`. 

* Changing the behavior of `max.k` in `knn.forecast.randomsearch.tuning()` when the current or previous default argument is passed to be more dynamic based on the length of the input series. Now `max.k` will be set to `min(floor((length(y.in)) * .4), length(y.in) - val.holdout.len - test.h)` if `NULL` or `NA` is passed. This change was made because the more arbitrary behavior in `knnwtsim 0.1.0` where `max.k` would be set to `min(floor((length(y.in)) * .4), 50)` if `NA` was passed could lead to errors.

* Error handling to throw errors or warnings for conflicting arguments, arguments outside reasonable ranges, and argument types differing from those listed in the help files has been added to all user facing functions in the package: `StMatrixCalc()`, `SpMatrixCalc()`, `SxMatrixCalc()`, `SwMatrixCalc()`, `knn.forecast()`,  `knn.forecast.randomsearch.tuning()`, and `knn.forecast.boot.intervals()`.

## Minor Changes

* A `min.k` argument has been added to `knn.forecast.randomsearch.tuning()`, which can be used set a floor for the minimum number of nearest neighbors to be proposed in any parameter set to be tested in tuning. By default `min.k = 1`, in line with previous behavior of the function.

* Added references to arXiv:2112.06266 to `DESCRIPTION` and help files as needed to provide more information on methodology. 



tests/testthat.R
library(testthat)
library(knnwtsim)

test_check("knnwtsim")


R/tuning.R

#' Tune \code{knn.forecast()} Hyperparameters with Random Search
#'
#' A simplistic automated hyperparameter tuning function which randomly
#' generates a grid of hyperparameter sets used to build corresponding \code{S_w} similarity matrices
#' which are used in \code{knn.forecast()} test against the last \code{test.h} points of \code{y.in} after
#' any \code{val.holdout.len} points are removed from the end of \code{y.in}. The best performing set of
#' parameters based on MAPE over over the forecast horizon of \code{test.h} points are returned as part of a list
#' alongside the 'optimum' weighted similarity matrix \code{Sw.opt}, the \code{Grid} of tested sets, and the MAPE
#' results. MAPE is the average of absolute percent errors for each point calculated as: \code{abs((test.actuals - test.forecast.i) / test.actuals) * 100}. Where \code{test.forecast.i}
#' and \code{test.actuals} are both numeric vectors.
#'
#'
#' @param grid.len integer value representing the number of hyperparameter sets to generate and test, must be \code{>= 1}.
#' @param St.in numeric and symmetric matrix of similarities, can be generated with \code{StMatrixCalc()}.
#' @param Sp.in numeric and symmetric matrix of similarities, can be generated with \code{SpMatrixCalc()}.
#' @param Sx.in numeric and symmetric matrix of similarities, can be generated with \code{SxMatrixCalc()}.
#' @param y.in numeric vector of the response series to be forecast.
#' @param test.h integer value representing the number of points in the test forecast horizon, must be \code{>= 1}.
#' @param max.k integer value representing the maximum value of k, \code{knn.forecast()} should use, will be set to \code{min(floor((length(y.in)) * .4), length(y.in) - val.holdout.len - test.h)} if \code{NULL} or \code{NA} is passed. Note this \code{NA} behavior differs from \code{knnwtsim} version 0.1.0.
#'  If a numeric value is passed it must be \code{>= 1}.
#' @param val.holdout.len integer value representing the number of observations at the end of the series to be removed in testing forecast if desired to leave a validation set after tuning, must be \code{>= 0}.
#' @param min.k integer value representing the minimum value of k, \code{knn.forecast()} should use, must be \code{>= 1}.
#'
#' @return list of the following components:
#' \describe{
#'           \item{weight.opt}{numeric vector of the 3 weights to generate \code{Sw.opt} in alpha, beta, gamma order which achieved the best performance in terms of MAPE.}
#'           \item{k.opt}{integer value of neighbors used in \code{knn.forecast()} which achieved the best performance in terms of MAPE.}
#'           \item{Sw.opt}{numeric matrix of similarities calculated using \code{S_w}, with the best performing set of hyperparameters.}
#'           \item{Test.MAPE}{numeric value of the MAPE result for the optimum hyperparamter set achieved on the test points.}
#'           \item{MAPE.all}{numeric vector of MAPE results, each observation corresponds to the row in \code{Grid} of the same index.}
#'           \item{Grid}{dataframe of all hyperparameter sets tested in the tuning.}
#' }
#'
#'
#'
#' @export
#'
#' @seealso
#' * Trupiano (2021) arXiv:2112.06266 for information on the formulation of \code{S_w}.
#' * [StMatrixCalc()] for information on the calculation of \code{S_t}.
#' * [SpMatrixCalc()] for information on the calculation of \code{S_p}.
#' * [SxMatrixCalc()] for information on the calculation of \code{S_x}.
#' * [knn.forecast()] for the function called to perform knn regression.
#' @md
#'
#'
#' @examples
#' data("simulation_master_list")
#' series.index <- 15
#' ex.series <- simulation_master_list[[series.index]]$series.lin.coef.chng.x
#'
#' df <- data.frame(ex.series)
#' # Generate vector of time orders
#' df$t <- c(1:nrow(df))
#' # Generate vector of periods
#' nperiods <- simulation_master_list[[series.index]]$seasonal.periods
#' df$p <- rep(1:nperiods, length.out = nrow(df))
#' # Pull corresponding exogenous predictor(s)
#' X <- as.matrix(simulation_master_list[[series.index]]$x.chng)
#'
#' St.ex <- StMatrixCalc(df$t)
#' Sp.ex <- SpMatrixCalc(df$p, nPeriods = nperiods)
#' Sx.ex <- SxMatrixCalc(X)
#'
#' tuning.test <- knn.forecast.randomsearch.tuning(
#'   grid.len = 10,
#'   y.in = ex.series,
#'   St.in = St.ex,
#'   Sp.in = Sp.ex,
#'   Sx.in = Sx.ex,
#'   test.h = 3,
#'   max.k = 10,
#'   val.holdout.len = 3
#' )
knn.forecast.randomsearch.tuning <- function(grid.len = 100,
                                             St.in,
                                             Sp.in,
                                             Sx.in,
                                             y.in,
                                             test.h = 1,
                                             max.k = NULL,
                                             val.holdout.len = 0,
                                             min.k = 1) {


  # integer argument type checks
  if ((!(is.vector(grid.len, mode = "numeric"))) |
    (!(identical(length(grid.len), 1L)))) {
    stop("grid.len should be an integer with length 1L")
  } else if (!(identical((grid.len %% 1), 0))) {
    warning("grid.len should be an integer
            , argument will be floored to nearest whole number")
    grid.len <- floor(grid.len)
  }

  if ((!(is.vector(test.h, mode = "numeric"))) |
    (!(identical(length(test.h), 1L)))) {
    stop("test.h should be an integer with length 1L")
  } else if (!(identical((test.h %% 1), 0))) {
    warning("test.h should be an integer
            , argument will be floored to nearest whole number")
    test.h <- floor(test.h)
  }

  if ((!(is.vector(val.holdout.len, mode = "numeric"))) |
    (!(identical(length(val.holdout.len), 1L)))) {
    stop("val.holdout.len should be an integer with length 1L")
  } else if (!(identical((val.holdout.len %% 1), 0))) {
    warning("val.holdout.len should be an integer
            , argument will be floored to nearest whole number")
    val.holdout.len <- floor(val.holdout.len)
  }

  if ((!(is.vector(min.k, mode = "numeric"))) |
    (!(identical(length(min.k), 1L)))) {
    stop("min.k should be an integer with length 1L")
  } else if (!(identical((min.k %% 1), 0))) {
    warning("min.k should be an integer
            , argument will be floored to nearest whole number")
    min.k <- floor(min.k)
  }

  # multi layered if statement is here because of the move to the recommended
  # default argument value without breaking calls on previous versions
  if (!identical(max.k, NULL)) {
    if (!identical(max.k, NA)) {
      if (((!(is.vector(max.k, mode = "numeric"))) |
        (!(identical(length(max.k), 1L))))) {
        stop("max.k should be an integer with length 1L, NULL, or NA")
      } else if (!(identical((max.k %% 1), 0))) {
        warning("max.k should be an integer
                , argument will be floored to nearest whole number")
        max.k <- floor(max.k)
      }
    }
  }

  # matrix argument checks
  if (!((is.matrix(St.in) & is.numeric(St.in))) |
    !((is.matrix(Sp.in) & is.numeric(Sp.in))) |
    !((is.matrix(Sx.in) & is.numeric(Sx.in)))) {
    stop("St.in, Sp.in, and Sx.in should be numeric matrices")
  } else if (!(isSymmetric.matrix(St.in)) |
    !(isSymmetric.matrix(Sp.in)) |
    !(isSymmetric.matrix(Sx.in))) {
    stop("St.in, Sp.in, and Sx.in should be a symmetric matrices")
  }

  if (!(identical(dim(St.in)[1], dim(Sp.in)[1])) |
    !(identical(dim(St.in)[1], dim(Sx.in)[1]))) {
    stop("Dimensions of St.in, Sp.in, and Sx.in are not all equal")
  }


  # response series checks
  if (!(is.vector(y.in, mode = "numeric"))) {
    stop("y.in should be a numeric vector")
  }

  n <- length(y.in)

  if (!(identical(dim(St.in)[1], n))) {
    stop("For tuning the length of y.in should match the number of rows of
         the similarity matrix inputs: St.in, Sp.in, Sx.in")
  }

  viable.neighbors.count <- n - val.holdout.len - test.h

  # set a maximum for k, multi layered if statement is here because of
  # the move to the recommended default argument value without breaking
  # calls on previous versions
  if (is.null(max.k)) {
    k.cap <- min(floor((length(y.in)) * .4), viable.neighbors.count)
  } else if ((is.na(max.k))) {
    k.cap <- min(floor((length(y.in)) * .4), viable.neighbors.count)
  } else {
    if (max.k > viable.neighbors.count) {
      warning(paste0("max.k is larger than the number of viable neighbors
                         given the test.h and val.holdout.len supplied, the
                         maximum number of viable neighbors will be used
                         instead: ", viable.neighbors.count))
    }
    k.cap <- min(max.k, viable.neighbors.count)
  }



  if (min.k > k.cap) {
    stop("min.k is greater than max.k")
  }

  # check if integer arguments are below minimum values
  if (any(c(grid.len, test.h, min.k, k.cap) < 1)){
    stop("None of the integer arguments: grid.len, test.h, min.k, and max.k
         can be < 1")
  }

  if (val.holdout.len < 0){
    stop("Integer argument val.holdout.len cannot be < 0")
  }


  # randomly propose sets of hyperparameters
  ks <- base::sample(min.k:k.cap, size = grid.len, replace = TRUE)
  alpha.0 <- stats::runif(grid.len, min = .001, max = 10000)
  beta.0 <- stats::runif(grid.len, min = .001, max = 10000)
  gamma.0 <- stats::runif(grid.len, min = .001, max = 10000)

  Theta <- data.frame(ks, alpha.0, beta.0, gamma.0)

  # normalize weights to sum to 1
  Theta$alpha <- Theta$alpha.0 / (Theta$alpha.0 + Theta$beta.0 + Theta$gamma.0)
  Theta$beta <- Theta$beta.0 / (Theta$alpha.0 + Theta$beta.0 + Theta$gamma.0)
  Theta$gamma <- Theta$gamma.0 / (Theta$alpha.0 + Theta$beta.0 + Theta$gamma.0)


  # remove any holdout points from y
  if (val.holdout.len > 0) {
    val.remove <- as.vector(c((n - val.holdout.len + 1):(n)))
    y.in <- y.in[-(val.remove)]
  }

  # create the test index based on test.h argument and
  # pull out actuals for comparison
  test.index <- c((length(y.in) - test.h + 1):length(y.in))
  test.actuals <- y.in[test.index]

  # set matrix to capture APE
  APE.h.i <- matrix(data = NA, nrow = nrow(Theta), ncol = length(test.index))

  # forecast loop to calculate test index APE for all Theta rows
  for (i in c(1:nrow(Theta))) {
    k.theta.i <- Theta[i, c("ks")]
    alpha.theta.i <- Theta[i, c("alpha")]
    beta.theta.i <- Theta[i, c("beta")]
    gamma.theta.i <- Theta[i, c("gamma")]

    # calculate the Sw matrix for the set of weights
    Sw.theta.i <- (alpha.theta.i * St.in + beta.theta.i * Sp.in
      + gamma.theta.i * Sx.in)

    # remove holdout points from Sw
    if (val.holdout.len > 0) {
      Sw.theta.i <- Sw.theta.i[-(val.remove), -(val.remove)]
    }

    # forecast using Sw.theta.i and compare to actuals
    test.forecast.i <- knn.forecast(
      Sim.Mat.in = Sw.theta.i,
      f.index.in = test.index,
      k.in = k.theta.i,
      y.in = y.in
    )
    APE.h.i[i, ] <- abs((test.actuals - test.forecast.i) / test.actuals) * 100
  }


  MAPE.h.i <- rowMeans(APE.h.i)

  # pull results for best performance in terms of MAPE over the forecast horizon
  Test.MAPE.opt <- MAPE.h.i[which.min(MAPE.h.i)]

  k.opt <- Theta[which.min(MAPE.h.i), c("ks")]
  alpha.opt <- Theta[which.min(MAPE.h.i), c("alpha")]
  beta.opt <- Theta[which.min(MAPE.h.i), c("beta")]
  gamma.opt <- Theta[which.min(MAPE.h.i), c("gamma")]

  weight.opt <- c(alpha.opt, beta.opt, gamma.opt)

  Sw.opt <- alpha.opt * St.in + beta.opt * Sp.in + gamma.opt * Sx.in


  # compile results for final list
  return.list <- list(
    weight.opt = weight.opt,
    k.opt = k.opt,
    Sw.opt = Sw.opt,
    Test.MAPE.opt = Test.MAPE.opt,
    MAPE.all = MAPE.h.i,
    Grid = Theta[, c("ks", "alpha", "beta", "gamma")]
  )

  return(return.list)
}
